### ISSUE TITLE
SuppressUnmanagedCodeSecurity

### ISSUE THREAD

**Issue by vbfox** [AArnott/pinvoke#202]
The `[SuppressUnmanagedCodeSecurity]` attribute is often placed on classes using PInvoke because otherwise any call to native dlls incur a performance cost.

Multiple questions here that I don't have an answer to :
- Is the performance problem still existing ?
- Is the attribute still useful ? The security model of .NET changed a lot since it's introduction and the idea of the CLR acting as a security barrier now seem pretty dead (Or any user-level VM by the way, kernel-level isolation as used in Chrome or Store Apps being much more secure).

I suspect that the performance cost is still there as CAS is deprecated since 4.0 but it's still partially present. If it's the case I think we should add the attribute everywhere as we want target relatively modern, 4.0+ apps & libraries.

**Comment by vbfox** [comment]
So for the first question, yes there is still a difference but while big in relative terms it's very small in absolute terms (At least on my 3GHz i7 CPU) :

SendMessage 100M calls **with** : 00:00:01.5157592
SendMessage 100M calls **without** : 00:00:03.0445422 _(+100%)_

So essentially the cost of not having `[SuppressUnmanagedCodeSecurity]` is **10ns** per call.
It's a big cost for APIs that need to be called in inner processing loops but for most of our current targets it's nothing (Anything that need a user -> kernel transition will take at least **100ns** to do it)

So except for SafeHandles where `SuppressUnmanagedCodeSecurity` is required on methods called by ReleaseHandle for coherency reasons (#203) I would suggest that we close the issue and open it again if anyone can show some real performance differences.

My test code : https://gist.github.com/vbfox/901b1ed05ff24a4ed9ec

**Comment by AArnott** [comment]
Thanks. I wasn't familiar with this attribute. I suspect it's only allowed to appear in assemblies in Full Trust. But our assembly already requires that by virtue of doing nothing but P/Invoke. Still, I suppose we have to consider security implications of applying this. So I agree, let's not do anything with this attribute unless there's a need to revisit.

**Comment by arlm** [comment]
Would this a good fit for the 2.0 milestone?

**Comment by AArnott** [comment]
Unless it's a breaking change, we could do it for a minor release.

**Comment by NN---** [comment]
@AArnott We have found that sometimes there is a huge overhead without SuppressUnmanagedCodeSecurity attribute.
Is it possible to generate a duplicate of each API with this attribute.
For instance

[DllImport("kernel32")] GetTickCount();
[SuppressUnmanagedCodeSecurity][DllImport("kernel32", EntryPoint = "GetTickCount")] GetTickCountSUCS(); 
Not sure about naming :)

This solves all problems.
Default is to have security, and if you need better performance call the other function.
What you think?

**Comment by AArnott** [comment]
It looks like the caller can apply the attribute to themselves and enjoy the perf gains. So no need for us to double up our API.

**Comment by NN---** [comment]
I think I missed that part in the documentation.
Can you point me to this?
Thanks.

**Comment by NN---** [comment]
https://msdn.microsoft.com/en-us/library/system.security.suppressunmanagedcodesecurityattribute(v=vs.110).aspx

This attribute is only effective when applied to PInvoke methods (or classes that contain PInvoke methods) or the definition of an interface through which interop calls will be made. It will be ignored in all other contexts.

For my understanding you cannot use this attribute on caller code.

**Comment by AArnott** [comment]
I was reading from https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2118-review-suppressunmanagedcodesecurityattribute-usage but it was a bit confusing and your MSDN link was much clearer, @NN---. I suspect you're right given the link you were reading from.

But rather than double up on all P/Invoke methods, given Code Access Security is a deprecated idea anyway, I'd be more in favor of just applying the attribute on all our classes. 

Are you (or anyone else) suffering from perf that this attribute would alleviate?

**Comment by NN---** [comment]
Agree classes is a better idea.
I just realized it after reading documentation once again and forgot to mention it.
Good catch.

For the question, the answer is yes.
For some reason there are machines where the difference is 100%.
Agree very odd and I suspect there is a problem somewhere.
Unfortunately I don't know who is responsible for such slower call, but adding attribute makes call faster.
I am investigating the cause, meanwhile I added attribute to all my pinvoke classes to reduce overhead.
